[#24] 강의 등록 시 Role check

kyeumd opened the pull request:
## 목표

강의 등록시 Role을 체크하여 강사가 아닌 경우 excpetion 발생

## 할 일

- [x] 강의 등록 Controller 에서 Role 체크
- [x] Role 체크에 대한 3가지 경우 테스트 코드 작성
> 1. 유효하지 않은 토큰
> 2. 유효하지 않은 권한
> 3. 정상 접근 권한

## 고려사항

- MockMVC과 RestAssured 함께 사용

Commit:
feat(#24) - 강의 등록 시 role check, 테스트 코드 작성

- Spring Security @PreAuthorize()를 통한 권한 체크
- 3가지 케이스 테스트 코드 작성
  1. 유효하지 않은 토큰
  2. 유효하지 않은 권한
  3. 정상 접근 권한
- 기존 multipart 테스트 코드 권한과 merge

chicori3 commented on src/test/java/com/flab/infrun/lecture/presentation/LectureControllerApiTest.java:
```diff
@@ -1,59 +1,129 @@
     @Autowired
     private MockMvc mockMvc;
-    private ObjectMapper mapper = new ObjectMapper();
+    private final ObjectMapper mapper = new ObjectMapper();
```
ObjectMapper에 따로 JavaTimeModule 같은 설정을 안한다면 @Autowired를 걸어도 괜찮은 것 같아요

chicori3 commented on src/test/java/com/flab/infrun/lecture/presentation/LectureControllerApiTest.java:
```diff
@@ -1,59 +1,129 @@
-        String lectureJson = mapper.writeValueAsString(lecture);
-        MockMultipartFile mockLecture = new MockMultipartFile("lecture", "lecture",
-            "application/json", lectureJson.getBytes(StandardCharsets.UTF_8));
+        mockMvc.perform(post("/lecture")
```
Arrange-Act-Assert나 Given-When-Then 패턴처럼 읽힐 수 있게 작성하면 더 테스트 코드를 읽기 수월할 것 같아요😄

chicori3 commented on src/main/java/com/flab/infrun/lecture/presentation/LectureController.java:
```diff
@@ -25,13 +26,11 @@ public class LectureController {
     public ResponseEntity<Response<Long>> registerLecture(
         @Valid @RequestPart("lecture") LectureRegisterRequest lecture,
         @RequestPart("file") List<MultipartFile> lectureVideoFile) {
-        //todo Role check (Teacher)
```
아직 제 PR이 승인되지 않아서 @CurrentUser를 쓸 수 없으시겠지만😂, 
Lecture를 생성할 때 강사의 ID(memberId) 정보는 request에 담겨있는 건가요?

@ResponseStatus로 어노테이션으로 응답 상태를 설정해주면 ResponseEntity를 사용하지 않고 Response만 반환해도 될 것 같습니다

kyeumd replied:
아 아직 Member ID를 넣는 부분이 todo 상태입니다! 시안님 PR 승인되면 @currentuser 정보로 member 정보를 가져올 생각입니다 🙃

ResponseEntity 로 반환 하는 것이 초기 프로젝트 convention  으로 생각하여 그대로 작성해왔는데 
covention을 ResponseStatus 어노테이션을 사용하는 것으로 변경하자는 말씀이신거죠? 
직접 컨트롤 하는 것 보다 어노테이션으로 사용하는 것에서 코드의 간결성에 대한 장점이 있다고 생각하셔서 해당 어노테이션을 사용하자고 권하시는걸까요?

시안님이 어떤 장점을 염두하시고 제안주신건지 궁금합니다 💁

chicori3 replied:
네 저도 처음에는 ResponseEntity<Response<T>> 식으로 작성하는 방법 밖에 몰랐고, 이 방법이 좋다고 생각했습니다.
아마 예전 멘토링 시간이었을텐데, @ResponseStatus에 대해 피드백을 받고 나서는 어노테이션을 활용하는 것도 괜찮은 것 같습니다.

ResponseEntity를 사용하는 이유는 Http Status를 함께 반환하기 위함인데 이는 @ResponseStatus 어노테이션도 마찬가지로 지원하며, ResponseEntity를 저희가 반복적으로 생성해서 핸들러의 반환값을 담지 않아도 되기 때문에 Response.success, Response.fail 만으로 처리할 수 있어서 더 간결하다는 생각이 들어요. 
물론 동겸님이 동의하지 않으신다면 제가 다시 맞출 수도 있는 부분 입니다!

kyeumd replied:
좋은 의견인것 같습니다. 제 생각도 두번 래핑하는 형태의 반환보다 Status를 @ResponseStatus로 반환하는 것이 더 깔끔할듯 합니다. 다음 PR에서 해당 부분 일괄 변경 작업 하겠습니다.
추후에 ResponseEntity 를 직접 컨트롤 해야하는경우(header setting 등)가 생긴다면 다시 의논해보면 좋을것 같습니다!

f-lab-michael commented on src/test/java/com/flab/infrun/lecture/presentation/LectureControllerApiTest.java:
```diff
@@ -1,59 +1,129 @@
     @Test
-    @DisplayName("Multipart 요청 파일과 request 객체가 함께 들어 온 경우 정상적인 처리 확인")
-    void lectureRegisterTest() throws Exception {
+    @DisplayName("강의 등록 시 Role check - 토큰 오류 (UN_AUTHORIZATION)")
```
권한 체크와 관련된 테스트중 실제 롤에 대한 테스트가 아닌 다른 부분(토큰의 유효성테스트등)은
비지니스로직의 테스트라기보다는 spring security 의 기능 테스트라고 볼수있습니다. 

공통적으로 해당 라이브러리의 동작을 확인하는 차원에서는 테스트이외에 매 기능마다 테스트하는것은 불필요하다고 생각합니다.

kyeumd replied:
End-to-End 테스트를 진행하고 싶어 Request부터 테스트를 진행하고자 하여
총 3가지 경우

1. JWT 토큰 자체가 잘못된 경우
2. 토큰은 문제 없고 @PreAuthorize 에서 Role의 권한 체크에서 통과하지 못한 경우(TEACHER 아닌 Role)
3. 성공한 경우

이렇게 총 3가지에 대해서 테스트 하고자 하여 작성했습니다. 
현재 비즈니스 자체에서 Role에 대한 처리가 없는 상황입니다.

Spring security 기능 테스트가 제가 하고자 하는 Role에 대한 테스트가 전부인 상황이라 3가지 상황에 대해서 테스트 코드를 작성하였습니다.

3가지 질문사항이 있습니다.

1. 비즈니스 안에서도 Role에 대한 권한을 체크하는 것이 좋을까요?
2. 라이브러리에 대한 테스트는 필요하지 않을까요?
3. 저는 3가지 기능의 테스트를 위해 3가지를 작성했는데 멘토님이 보시기에 어떤 부분이 불필요하게 중복된다고 하시는것인가요?

f-lab-michael replied:
end-to-end test 를 진행하시면 3가지중 3번만 진행해도 충분합니다.
3가지 질문사항
1. 어노테이션만으로 충분합니다.
2. 일반적으로 라이브러리 동작자체를 테스트 하진 않습니다. 그건 해당 라이브러리에서 이미 테스트된상태로 가정합니다.
3. jwt 토큰이 잘못된 케이스에 대한 테스트는 security 자체의 기능을 테스트 하는것이기 떄문에 1회만 하면되고, 구현하는 api 마다 테스트 할 필요가 없다는 얘기입니다.

kyeumd replied:
감사합니다 아직 라이브러리 사용에 익숙치 않다보니 제가 사용을 제대로 한지가 맞는지 의심스러워서 테스트해보게 되는것 같습니다 😅
테스트 부분 좀 더 간결하게 수정하고 재요청 드리겠습니다!

f-lab-michael commented on src/test/java/com/flab/infrun/lecture/presentation/LectureControllerApiTest.java:
```diff
@@ -1,59 +1,129 @@
+    }
+
+    private String loginSuccess_returnToken(LoginRequest loginRequest) {
+        var response = RestAssured.given()
```
var 사용시 어떤 장단점이 있을까요?
어떤 기준으로 var를 사용하고 계신가요?

kyeumd replied:
테스트코드에서 JWT 토큰에 대해 returnToken 하는 부분을 시안님이 작성한 코드를 보수적으로 copy 해오는 과정에서 해당 유형이 var임을 인지하지 못하고 그냥 사용했습니다. 😢
저는 개인적으로 Java의 var 타입의 유연함이 주는 장점보다 명시적인 타입으로 선언하여 명확하게 하는 방법을 선호합니다.
해당 반환타입은 명시적으로 변경하겠습니다!

chicori3 committed:
[#5] 장바구니 기능 구현 (#36)

* feat(#5) - 장바구니 도메인 작성

* feat(#5) - 장바구니 영속성 계층 작성

* feat(#5) - 장바구니 추가 서비스 작성

* test(#5) - 장바구니 추가 서비스 테스트 작성, 네이밍 수정

* feat(#5) - 장바구니 제거 도메인 로직 작성

* feat(#5) - 장바구니 제거 서비스 작성

* test(#5) - 장바구니 제거 서비스 테스트 작성, 네이밍 수정

* feat(#5) - 장바구니 표현 계층 작성

* refactor(#7) - SecurityContextHolder의 유저 정보를 받기 위한 코드 추가

* chore(#5) - 시큐리티 테스트를 위한 의존성 추가

* test(#5) - 장바구니 표현 계층 테스트 작성

* refactor(#5) - 불필요한 코드 제거, 네이밍 수정, 장바구니 관련 상위 예외 작성

* feat(#5) - 장바구니 조회 서비스 작성

* test(#5) - 장바구니 조회 서비스 테스트 작성

* feat(#5) - 장바구니 조회 응답 객체 작성

* refactor(#5) - 장바구니 추가, 삭제 시에는 현재 장바구니에 담긴 강의 ID만 반환하도록 변경

* feat(#37) - CartItems 일급 컬렉션으로 변경, 수강바구니에 없는 강의 삭제 요청 시 예외 발생 로직 추가

* refactor(#37) - 불필요한 DeleteCartItemRequest 삭제, 유효성 검사 추가, 테스트 작성

chicori3 commented:
RestAssured와 MockMvc를 같이 쓰셨다고 했는데 둘을 같이 쓴 이유를 알 수 있을까요?

chicori3 committed:
feat(#43) - Querydsl 설정 추가 (#44)

chicori3 committed:
[#45] 장바구니 리팩토링 (#46)

refactor(#45) - Lecture 객체의 Member 조회 추가, 코드 수정 (#46)

chicori3 committed:
[#37] 쿠폰 조회 기능 추가, 리팩토링 (#39)

* refactor(#37) - 코드 에러 수정, 연관관계 ID로 변경

* refactor(#37) - 예외 시 ErrorCode를 넘기지 않도록 변경

* feat(#37) - 쿠폰 사용 불가 예외 추가

* feat(#37) - 쿠폰 조회 기능 구현

* test(#37) - 쿠폰 조회 기능 테스트 작성

kyeumd commented:
시안님이 작성하신 JWT 토큰을 반환하는 부분을 최대한 살려서 가져오다보니 RestAssured를 추가로 사용하게되었습니다.

Commit:
refactor(#24) - feedback 리팩토링 작업

- Autowired 사용
- Given-When-Then 형태의 패턴으로 변경
- @CurrentUser를 통한 Lecture 생성 시 Member 주입
- 권한 체크 관련 불필요한 테스트 메서드 삭제
- var 타입 반환 수정
